src/core/system_parameters/beidou_cnav2_navigation_message.cc

```diff
@@ -0,0 +1,527 @@
+        {
+            frame_ID = 5;
+        }
+    else
+        {
+            LOG(WARNING) << "beidou cnav2: Invalid Satellite Slot Number";
+            frame_ID = 0;
+        }
+
+    return frame_ID;
+}
+
+
+int beidou_cnav2_Navigation_Message::string_decoder(std::string frame_string)
+{
+    int J = 0;
+    d_string_ID = 0;
+    d_frame_ID = 0;
+
+    // Unpack bytes to bits
+    std::bitset<BEIDOU_CNAV2_STRING_BITS> string_bits(frame_string);
+
+    // Perform data verification and exit code if error in bit sequence
+    flag_CRC_test = CRC_test(string_bits);
+    if (flag_CRC_test == false)
+        return 0;
+
+    // Decode all 15 string messages
+    d_string_ID = static_cast<unsigned int>(read_navigation_unsigned(string_bits, STRING_ID));
+    switch (d_string_ID)
+        {
+    // These should be changed according to the format outlined in the BEIDOU_B2a_CA.h
+
+        case 1:
+            //--- It is Type 10 -----------------------------------------------
+        	/*
+
+
+			*/
+            break;
+
+        case 2:
+            //--- It is Type 11 -----------------------------------------------
+        	/*
+            if (flag_ephemeris_str_1 == true)
+                {
+
+                }
+            */
+            break;
+
+        case 3:
+            // --- It is Type 30 ----------------------------------------------
+        	/*
+            if (flag_ephemeris_str_2 == true)
+                {
+
+                }
+            */
+            break;
+
+        case 4:
+            // --- It is Type 31 ----------------------------------------------
+        	/*
+            if (flag_ephemeris_str_3 == true)
+                {
+
+                }
+            */
+            break;
+
+        case 5:
+            // --- It is Type 32 ----------------------------------------------
+        	/*
+            if (flag_ephemeris_str_4 == true)
+                {
+
+                }
+            */
+            break;
+
+        case 6:
+            // --- It is Type 33 ----------------------------------------------
+        	/*
+
+            */
+            break;
+
+        case 7:
+            // --- It is Type 34 ----------------------------------------------
+        	/*
+            if (flag_almanac_str_6 == true)
+                {
+
+                }
+            */
+            break;
+
+        case 8:
+            // --- It is Type 40 ----------------------------------------------
+        	/*
+
+            */
+            break;
+
+
+        default:
+            LOG(INFO) << "beidou cnav2: Invalid String ID of received. Received " << d_string_ID
+                      << ", but acceptable range is from 10 11 30 31 32 33 34 40";
+
+            break;
+        }  // switch string ID
+
+    return d_string_ID;
+}
+
+
+Beidou_Cnav2_Ephemeris beidou_cnav2_Navigation_Message::get_ephemeris()
+{
+    return cnav2_ephemeris;
+}
+
+/*
+Beidou_Cnav2_Utc_Model beidou_cnav2_Navigation_Message::get_utc_model()
```

Comment:
time parameters inside the navigation message should go inside this field. Also, other signals provide iono parameters and they used the `iono_model` type. Look at what GPS does but I think Beidou is very similar. This might be needed. Ask questions if not sure